Fix: Consider input defined unbacked during inductor codegen for runtime asserts

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152231

So when we use mark_unbacked the graph will have an unbacked inputs symInt. Right now, 
deferred runtime assertions that uses those  is never generated.

This PR changes that, such that in the forward graph we consider those and generate the corresponding 
runtime assertions of them. We still ignore them for backward which is not ideal

The way we generate runtime assertion is by emitting them when all the defined unbacked symbols used 
in them are seen. 

We previously skipped placeholder, because for backward we have a wacky approach were we 
ignore input defined unbacked symbols and assumes assertions that uses them are already emitted
in forward and we try to emit all other runtime assertions again. see [Note [Backwards runtime asserts]

Doing that we ends up only emitting the runtime assertions that depends on things defined solely in backward, but we could miss checks that spans inputs defined in both backward and forward, i.e one symbol defined in forward passed as input to backward., and another that is defined in backward.) .This is not ideal an ideal approach could be something like this https://github.com/pytorch/pytorch/pull/151919 but it require more work .


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
consisder input defined unbacked durign inductor codegen for runtime asserts

[ghstack-poisoned]

laithsakka commented:
The failing test is not harmlful will fix before landing it.

laithsakka committed:
Update on "Fix: Consider input defined unbacked during inductor codegen for runtime asserts"


So when we use mark_unbacked the graph forward graph will have an unbacked inputs symInt. 
right now the runtime assertions that uses that is never codegen.

The way we generate runtime assertion is by emitting them when all the defined unbacked symbols used 
in them are emitted. 

We previously skipped placeholder, because for backward we have a wacky approach were we 
skip input defined unbacked symbols and assumes assertions that uses them are already emitted
in forward and we try to emit all runtime assertions again otherwise. see [Note [Backwards runtime asserts]

with that we would end up only emitting those that depends on things defined soleley in backward, but we could miss checks that spans inputs defined in both backward and forward, i.e one symbol defined in forward passed as input to backward., and another that is defined in backward.) .This is not ideal an ideal approach could be something like this 
https://github.com/pytorch/pytorch/pull/151919 but it require more work .

That said, there is no reason though to skip place holders for forward so for now doing that. 



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

aorenste commented on test/test_dynamic_shapes.py:
```diff
@@ -3033,6 +3033,23 @@ def func(a, b):
         with self.assertRaises(RuntimeError):
             func(torch.tensor([100]), torch.tensor([1]))
 
+    @skipIfTorchDynamo("mark_unbacked is not tracable")
```
nit: traceable

aorenste commented on torch/_inductor/graph.py:
```diff
@@ -1772,6 +1772,17 @@ def debug(msg: str) -> None:
 
+        shape_env = V.graph.sizevars.shape_env
+
+        # An input can an unbacked symint i.e.: when mark_unabcked is ued symbol
```
```suggestion
        # An input can be an unbacked symint i.e.: when mark_unbacked is used
```

aorenste commented on torch/_inductor/graph.py:
```diff
@@ -1783,96 +1794,98 @@ def format_new_defs() -> str:
-                f"fx node is: {n.format_node()}\n"
-                f"new operations are:\n\n{format_new_defs()}"
-            )
+        # We do not skip unbacked symints that are input for backward see the note bellow.
```
```suggestion
        # We do not skip unbacked symints that are input for backward see the note below.
```

aorenste reviewed (approved):
This looks reasonable to me... just a few spelling nits

laithsakka commented:
Fix typos in the summary

bobrenjc93 commented:
s/Tight now/Right now

in PR summary

laithsakka committed:
Update on "Fix: Consider input defined unbacked during inductor codegen for runtime asserts"


So when we use mark_unbacked the graph will have an unbacked inputs symInt. Right now, 
deferred runtime assertions that uses those  is never generated.

This PR changes that, such that in the forward graph we consider those and generate the corresponding 
runtime assertions of them. We still ignore them for backward which is not ideal

The way we generate runtime assertion is by emitting them when all the defined unbacked symbols used 
in them are seen. 

We previously skipped placeholder, because for backward we have a wacky approach were we 
ignore input defined unbacked symbols and assumes assertions that uses them are already emitted
in forward and we try to emit all other runtime assertions again. see [Note [Backwards runtime asserts]

Doing that we ends up only emitting the runtime assertions that depends on things defined solely in backward, but we could miss checks that spans inputs defined in both backward and forward, i.e one symbol defined in forward passed as input to backward., and another that is defined in backward.) .This is not ideal an ideal approach could be something like this https://github.com/pytorch/pytorch/pull/151919 but it require more work .


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge
